Replace google drive with Firebase as collaboration backend #9003

jamessmith126 opened the pull request:
<!--- Replace google drive with Firebase as collaboration backend #9003 -->
## Description

I have create a new real time manager class named FirebaseRealtimeManager.
I have renamed current RealtimeManager to GoogleDriveRealtimeManager.
About speaking of FirebaseRealtimeManager, I have initiated it with same interface and event handlers should be enough.
I have added new packages("firebase": "3.5.1", "firebase-api": "1.0.0", "firebase-client": "0.1.1",) in ./client/package.json.
I have created RealtimeManagerFactory for Firebase and GoogleDrive.
## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

<!--- If it fixes an open issue, please link to the issue here. -->
## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->

<!--- Include details of your testing environment, and the tests you ran to -->

<!--- see how your change affects other areas of the code, etc. -->
## Screenshots (if appropriate):
## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

goldendev601 committed:
Update package.json

goldendev601 committed:
Update and rename realtimemanager.coffee to googledriverealtimemanager.coffee

goldendev601 committed:
Update collaborationcontroller.coffee

goldendev601 committed:
Update collaborationcontroller.coffee

goldendev601 committed:
Create firebaserealtimemanager.coffee

goldendev601 committed:
Update firebaserealtimemanager.coffee

goldendev601 committed:
Update googleapiclient.coffee

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/koding/koding?pullRequest=9389) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/koding/koding?pullRequest=9389) before we can accept your contribution.<br/>

szkl commented on package.json:
```diff
@@ -48,6 +48,9 @@
     "faker": "1.1.0",
+    "firebase": "3.5.1",
+    "firebase-api": "1.0.0",
+    "firebase-client": "0.1.1",
```
These packages should be added into dependencies list in client/package.json file.

szkl commented on client/ide/lib/googleapiclient.coffee:
```diff
@@ -3,6 +3,8 @@ KDObject = kd.Object
 KodingAppsController = require 'app/kodingappscontroller'
 IDEMetrics = require './idemetrics'
 $ = require 'jquery'
+GOOGLE_OATH_SCOPES = 'email, https://www.googleapis.com/auth/drive.metadata.readonly'
```
Let's separate variable declaration statement from require calls.

szkl reviewed (changes requested):
Firebase Realtime Manager should not have any reference to Google API. Underlying realtime event mechanisms are still based on Google API here.

szkl commented:
All so, it'd make it more readable and understandable if you explain your changes in detail in respective commit messages. "Created a file" and then "Updated a file" are not helpful. You can amend last commit if successive commits doesn't make a huge change that requires you to explain in the commit description.

goldendev601 committed:
Update googleapiclient.coffee

goldendev601 committed:
Router: use provided anchor info for taking action on Home app nav

hakankaradis commented:
Create a factory that is called RealtimeManagerFactory.
CollaborationController will use this factory to create realtimemanager instances. 
These realtimemanager instances will have same generic methods and event handlers.

For instance, when we want to create FirebaseRealtimeManager instance.
we will be able to create that instance like this `RealtimeManagerFactory.get({ type: 'FIREBASE' })` or `RealtimeManagerFactory.get({ type: 'GOOGLE_DRIVE' })`

So current existing googledriverealtimemanager class will be created by using this factory and same for firebaserealtimemanager.

https://en.wikipedia.org/wiki/Factory_method_pattern
/cc: @szkl

goldendev601 committed:
Create a FirebaseRealtimeManager

This is new real time manager class named called FirebaseRealtimeManager.
It uses same interface and event handler with GoogleDirveRealtimeManager.

goldendev601 committed:
Create FirebaseRealtimeManager

This is new real time manager class named called FirebaseRealtimeManager.
It uses same interface and event handler with GoogleDirveRealtimeManager.

goldendev601 committed:
Update CollaborationController

CollaborationController use this RealtimeManagerFactory to create realtimemanager instances.
like this:
   RealtimeManagerFactory.get({ type: 'FIREBASE' }) 
   RealtimeManagerFactory.get({ type: 'GOOGLE_DRIVE' })

goldendev601 committed:
Create RealtimeManagerFactory

Factory for FirebaseRealtimeManager and GoogleDriveRealtimeManager

goldendev601 committed:
Update package.json

goldendev601 committed:
Update googleapiclient.coffee

goldendev601 committed:
Update: include all event handlers for gapi and firebase

gokmen commented on client/ide/lib/realtimemanagerfactory.coffee:
```diff
@@ -8,8 +8,363 @@ module.exports = class RealtimeManagerFactory extends KDObject
+    if type is 'FIREBASE'
+      firebase.put(options).execute (file) =>
+      callback null, file
+    else if type is 'GOOGLE_DRIVE'
```
these shouldn't be handled like this. you need to implement an interface which will be base for each realtime provider and after that you need to implement them in place in their classes;

Here is an example;

### Interface
```coffee
class RealTimeInterface
  
  createFile: (options, callback) ->
    callback { message: 'createFile: Not implemented' }

  getFile: (options, callback) ->
    callback { message: 'getFile: Not implemented' }
...
```

### GoogleDrive Implementation
```coffee
class GoogleDriveInterface extends RealTimeInterface
  
  createFile: (options, callback) ->
    options      =
      resource   :
        mimeType : 'application/vnd.google-apps.drive-sdk'
        title    : title
    gapi.client.drive.files.insert(options).execute (file) ->
      callback null, file

  getFile: (options, callback) ->
    fileId = options.id
    gapi.client.drive.files.get({ fileId }).execute (file) =>
      @emit 'FileFetched', file
      @loadFile { id: file.id }, callback
...
```

### Firebase Implementation
```coffee
class FirebaseInterface extends RealTimeInterface
  
  createFile: (options, callback) ->
    options      =
      resource   :
        mimeType : 'application/vnd.google-apps.drive-sdk'
        title    : title 
    firebase.put(options).execute (file) ->
      callback null, file

  getFile: (options, callback) ->
    fileId = options.id
    firebase.get({ fileId }).execute (file) => 
      @emit 'FileFetched', file
      @loadFile { id: file.id }, callback...
...
```

you need to change all other methods has `if` blocks in it for provider based implementations.

jamessmith126 replied:
I have implemented interface.

szkl replied:
Factory class is not an interface type. It doesn't need to declare methods.

You should've created a new class named `RealtimeManagerInterface`. `RealtimeManagerFactory` should have only a getter instantiating new instances.

gokmen commented on client/ide/lib/realtimemanagerfactory.coffee:
```diff
@@ -8,8 +8,363 @@ module.exports = class RealtimeManagerFactory extends KDObject
+      if type is 'FIREBASE'
+        firebase.remove() =>
+        callback null
+        @emit 'FileDeleted'  unless preventEvent
```
wrong indentation.

gokmen commented on client/ide/lib/realtimemanagerfactory.coffee:
```diff
@@ -8,8 +8,363 @@ module.exports = class RealtimeManagerFactory extends KDObject
+    else if type is 'GOOGLE_DRIVE'
+      gapi.client.drive.files.insert(options).execute (file) =>
+      callback null, file
+      @emit 'FileCreated', file  unless preventEvent
```
wrong indentation.

gokmen commented on client/ide/lib/realtimemanagerfactory.coffee:
```diff
@@ -8,8 +8,363 @@ module.exports = class RealtimeManagerFactory extends KDObject
+        title    : title
+    if type is 'FIREBASE'
+      firebase.put(options).execute (file) =>
+      callback null, file
```
wrong indentation and fat arrow is not necessary.

jamessmith126 replied:
I have updated.

gokmen commented on client/ide/lib/realtimemanagerfactory.coffee:
```diff
@@ -8,8 +8,363 @@ module.exports = class RealtimeManagerFactory extends KDObject
+        when ErrorType.TOKEN_REFRESH_REQUIRED then 'ErrorRealtimeTokenExpired'
+        else 'ErrorHappened'
+
+      @emit eventName, error
```
this method has too many internal method definitions here.

gokmen commented on client/ide/lib/realtimemanagerfactory.coffee:
```diff
@@ -8,8 +8,363 @@ module.exports = class RealtimeManagerFactory extends KDObject
+    return data
+  
+
+  retrieveAllFiles: () ->
```
no need for parentheses.

gokmen commented on client/ide/lib/realtimemanagerfactory.coffee:
```diff
@@ -8,8 +8,363 @@ module.exports = class RealtimeManagerFactory extends KDObject
+
+  retrieveAllFiles: () ->
+    retrievePageOfFiles = @(request) ->
+        request.execute(@(resp) ->
```
wrong indentation.

gokmen commented on client/ide/lib/realtimemanagerfactory.coffee:
```diff
@@ -8,8 +8,363 @@ module.exports = class RealtimeManagerFactory extends KDObject
+  
+
+  retrieveAllFiles: () ->
+    retrievePageOfFiles = @(request) ->
```
this usage is not looking good at all.

goldendev601 committed:
Update firebaserealtimemanager.coffee

goldendev601 committed:
Update firebaserealtimemanager.coffee

goldendev601 committed:
Update googledriverealtimemanager.coffee

goldendev601 committed:
Update realtimemanagerfactory.coffee

goldendev601 committed:
Update realtimemanagerfactory.coffee

szkl commented:
I have to repeat this.

> it'd make it more readable and understandable if you explain your changes in detail in respective commit messages. "Created a file" and then "Updated a file" are not helpful. You can amend last commit if successive commits doesn't make a huge change that requires you to explain in the commit description.

szkl commented on client/ide/lib/realtimemanagerfactory.coffee:
```diff
@@ -0,0 +1,152 @@
+        realtime = new FirebaseRealtimeManager
+      else if type is 'GOOGLE_DRIVE'
+        realtime = new GoogleDriveRealtimeManager
+      return realtime
```
Define a map of name => constructor in class level.

```coffeescript
class RealtimeManagerFactory

typeMap =
  FIREBASE: require './firebaserealtimemanager'
  GOOGLE_DRIVE: require './googledriverealtimemanager'

  get: (type) -> new typeMap[type]
```

szkl commented on client/ide/lib/firebaserealtimemanager.coffee:
```diff
@@ -0,0 +1,373 @@
+
+  constructor: (@KodingApp) ->
+    super options, data
+    @firebase = new Firebase("https://KodingApp.firebaseio.com")
```
`Firebase` is not defined.

szkl replied:
`firebase` references in this file are not defined too.

firebase.database namespace doesn't have any named `Event`. https://firebase.google.com/docs/reference/js/

Method names are still derived from Google Drive API.

szkl replied:
Event subscription codes are invalid and looks like just copied pasted from Drive Realtime Manager class.

goldendev601 committed:
Update firebaserealtimemanager.coffee

goldendev601 committed:
Update firebaserealtimemanager.coffee

goldendev601 committed:
Update firebaserealtimemanager.coffee

goldendev601 committed:
Update realtimemanagerfactory.coffee

goldendev601 committed:
Update: configuration and connection to firebase

goldendev601 committed:
modify line 500 and 1132, delete some part

goldendev601 committed:
delete some no need part

szkl commented on client/ide/lib/firebaserealtimemanager.coffee:
```diff
@@ -4,33 +4,40 @@ IDEMetrics       = require './idemetrics'
+      databaseURL: "https://koding-cff25.firebaseio.com/",
+      storageBucket: "gs://koding-cff25.appspot.com"
+    };
+    firebase.initializeApp(config)
```
Config object should be designated in client runtime options. Check out default configuration modules.

[main.default.coffee](https://github.com/koding/koding/blob/master/config/main.default.coffee)
[credentials.default.coffee](https://github.com/koding/koding/blob/master/config/credentials.default.coffee)

[config/generateRuntimeConfig](https://github.com/koding/koding/blob/master/config/generateRuntimeConfig.coffee) module generates configuration object for client application.

szkl commented on client/ide/lib/firebaserealtimemanager.coffee:
```diff
@@ -4,33 +4,40 @@ IDEMetrics       = require './idemetrics'
+  
   ###*
-   * Create file.
+   * Create file to google drive.
```
This is `client/ide/lib/firebaserealtimemanager.coffee`.
